---
title: "Assignment 1"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---
# Data analysis
## Data set series: GSE198919
## Dataset title: Transcriptome and methylation analysis of JMML patients compare to healthy prenatal and postnatal samples.

JMML (Juvenile myelomonocytic leukemia) is a leukemia that only develops in young children and is thought to have a prenatal initiation (Cavé H et al). This study is about RNAseq expression and DNA methylation experiment between JMML patients, healthy prenatal and postnatal samples (Cavé H et al). The researchers studied the transcriptome of hematopoietic stem and progenitor cells(HSPC) sorted from sporadic JMML patients, healthy prenatal samples and from healthy age matched donors (Cavé H et al). Bulk transcriptome of sorted HSPC reveals that some JMML samples cluster with prenatal samples whereas other from a distinct cluster apart from any healthy samples (Cavé H et al). Methylation profile on bulk mononucleated cell on theses JMML patients, 2 healthy postnatal and 2 healthy prenatal samples is also investigated (Cavé H et al). The results show a global hypermethylation in JMML samples compared to healthy samples and a specific JMML group with a hypermethylated profile compared to all JMML samples (Cavé H et al).

```{r Download required packages}
if(!requireNamespace("knitr", quietly=TRUE)) {
 install.packages("knitr")
}
if(!requireNamespace("edgeR", quietly=TRUE)) {
  install.packages("edgeR")
}
if(!requireNamespace("BiocManager", quietly=TRUE)) {
  install.packages("BiocManager")
}
if(!requireNamespace("GEOmetadb", quietly=TRUE)) {
  BiocManager::install("GEOmetadb")
}
if(!requireNamespace("biomaRt", quietly=TRUE)) {
  BiocManager::install("biomaRt")
}
```
```{r library packages needed}
library(knitr)
library(GEOmetadb)
library(biomaRt)
library(edgeR)
```
```{r Get dataset and meta file, message=FALSE, warning=TRUE}
dataset_name <- "GSE198919"
gse <- getGEO(dataset_name, GSEMatrix = FALSE)
kable(data.frame(head(Meta(gse))), format="html", caption = paste0("Metadata of ", dataset_name))
```
```{r Download supplementary files, message=FALSE, warning=FALSE}
sfiles <- getGEOSuppFiles(dataset_name)
(fnames <- rownames(sfiles))
```
```{r Unzip RNA raw counts file and read the file}
if(!file.exists(sub(".gz", "", fnames[4]))) {
  gunzip(fnames[4], remove=FALSE)
}

raw_counts <- read.delim(sub(".gz", "", fnames[4]), header=TRUE, check.names=FALSE)
```
```{r Show some gene names}
raw_counts$Gene[10:20]
raw_counts$Gene[60:70]
```
```{r Get the dataset of ensembl}
ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart=ensembl)
```
In the dataset, most of gene names were in HUGO symbol.
```{r Filter the gene names with HUGO symbol}
raw_counts_gene_conversion <- getBM(attributes=c("hgnc_symbol", "hgnc_symbol"),
      filters=c("hgnc_symbol"),
      values=raw_counts$Gene,
      mart=ensembl)
```
However, there were some gene names were in HUGO symbols.
```{r Calculate how many genes were not in HUGO symbols}
nrow(raw_counts) - nrow(raw_counts_gene_conversion)
```
Convert the raw_counts data frame into matrix and merge with mapped gene names. Filter the rows with a valid HUGO symbol.
```{r Merge}
raw_counts_matrix <- as.matrix(raw_counts)
rownames(raw_counts_matrix) <- raw_counts$Gene
counts_annot <- merge(raw_counts_gene_conversion, raw_counts_matrix, by.x=1, by.y=0, all.y=TRUE)
counts_annot <- counts_annot[which(!is.na(counts_annot$hgnc_symbol.1)),]
head(counts_annot)
```
The researches named the samples with JMML as LMMJ(leucémie myélomonocytaire juvénile). Other samples are healthy.
```{r Cell type}
cell_type <- c()
for (i in colnames(counts_annot)[4:ncol(counts_annot)]) {
  if(grepl("LMMJ", i)) {
    cell_type <- c(cell_type, "JMML")
  } else {
    cell_type <- c(cell_type, "Healthy")
  }
}
```
```{r convert into matrix}
counts_annot_matrix <- as.matrix(sapply(counts_annot[,4:ncol(counts_annot)], as.numeric))

rownames(counts_annot_matrix) <- counts_annot$hgnc_symbol
```
```{r Calculate cpm}
cpms <- cpm(counts_annot_matrix[,4:ncol(counts_annot_matrix)])
keep <- rowSums(cpms>1) >= 10
counts_annot_filtered_matrix <- counts_annot_matrix[keep,]
dim(counts_annot_filtered_matrix)
```
```{r plot CPM data, warning=FALSE}
cpm_log_data <- log2(cpm(counts_annot_filtered_matrix))
boxplot(cpm_log_data, xlab="Samples", ylab="log2 CPM",
        las=2, cex=0.5, cex.lab=0.5,
        cex.axis=0.5, main="JMML RNASeq Samples")
abline(h=median(apply(cpm_log_data, 2, median)),
       col="green", lwd=0.6,lty="dashed")
```
```{r plot density graph of RNAseq}
# Outlier: LMMJ-10-CMP
density_data <- apply(log2(cpm(counts_annot_filtered_matrix[,1:ncol(counts_annot_filtered_matrix)])), 2, density)
xlim <- 0
ylim <- 0
for (i in 1:length(density_data)) {
  xlim <- range(c(xlim, density_data[[i]]$x))
  ylim <- range(c(ylim, density_data[[i]]$y))
}

cols <- rainbow(length(density_data))
ltys <- rep(1, length(density_data))

plot(density_data[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM", main="", cex.lab=TRUE)
for (i in 1:length(density_data)) {
  lines(density_data[[i]], col=cols[i])
}

legend("topright", colnames(counts_annot_filtered_matrix), 
       col=cols, lty=ltys, ncol=3, cex=0.3,
       border="blue", text.col="green4",
       merge=TRUE, bg="gray90")

```
As we can see in the plot, LMMJ-10-CMP stands out from the dataset significantly. Therefore, LMMJ-10-CMP is an outlier here.


```{r Calculate normalize factors}
d <- DGEList(counts=counts_annot_filtered_matrix, group=cell_type)
d <- calcNormFactors(d)
plotMDS(d, labels=colnames(counts_annot_filtered_matrix),
        col=c("red", "blue")[factor(cell_type)], cex=0.7)
```

# Interpretation

Control condition is healthy and test condition is JMML. This dataset is interesting for me because we can compare RNA expression between among JMML patients, healthy prenatal sample, healthy postnatal sample. There were expression values that were not unique for specific genes. I discarded these genes in the final data analysis. There were expression values that could not be mapped to current HUGO symbols. I discarded these genes because it is too difficult to search.